include/boost/compute/random/discrete_distribution.hpp

```diff
@@ -42,8 +45,8 @@ class discrete_distribution
     /// the range [\p first, \p last)
     template<class InputIterator>
     discrete_distribution(InputIterator first, InputIterator last)
-        : m_n(std::distance(first, last)),
-          m_probabilities(std::distance(first, last))
+        : m_n((std::max)(size_t(1), static_cast<size_t>(std::distance(first, last)))),
+          m_probabilities((std::max)(size_t(1), static_cast<size_t>(std::distance(first, last))))
```

Comment:
I think we still need to handle the case where `first == last` better. Looking at the documentation for [std::discrete_distribution](http://en.cppreference.com/w/cpp/numeric/random/discrete_distribution/discrete_distribution), when the input range is empty (or calling the default constructor), we should just initialize with a single probability of `1`. But this is currently not implemented, and we don't have to do it in this pull request. I'll open a separate bug for this.